[openssl] openssl/1.1.1j Fails to build first time (once) when using cmake-conan with "cannot open input file 'apps\app_rand.obj'"

### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **openssl/1.1.1j**
  * Operating System+version: **Windows 10**
  * Compiler+version: **Visual Studio 2017**
  * Conan version: **conan 1.32.1**
  * Python version: **Python 3.7.4**

### Steps to reproduce (Include if Applicable)

Let's say we have `conan-bug` directory, where inside we have:

**profile.txt**:
```
include(default)
[settings]
compiler="Visual Studio"
compiler.version=15
arch=x86
arch_build=x86
```

**conanfile.txt**:
```
[requires]
openssl/1.1.1j@

[options]
openssl:shared=True

[generators]
cmake_find_package
cmake_paths
```

**CMakeLists.txt**:
```
cmake_minimum_required(VERSION 3.16)

project(ConanBug)

list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake-conan-0.15")
include(conan)

add_executable(${PROJECT_NAME})

conan_cmake_run(
    BUILD missing
    BUILD_TYPE "${CMAKE_BUILD_TYPE}"
    CONANFILE conanfile.txt
    INSTALL_FOLDER "${PROJECT_SOURCE_DIR}/build"
    PROFILE "${PROJECT_SOURCE_DIR}/profile.txt"
    SETTINGS "build_type=${CMAKE_BUILD_TYPE}"
)
include(${PROJECT_SOURCE_DIR}/build/conan_paths.cmake)
```

**cmake-conan-0.15** directory from:
https://github.com/conan-io/cmake-conan/releases/tag/v0.15

Now, in `cmd.exe` run:

```
conan.exe remove openssl --force
"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Auxiliary\Build\vcvars32.bat"
"C:\Program Files\CMake\bin\cmake.exe" -G Ninja -DCMAKE_BUILD_TYPE=Debug -B build .
```

Output with failure:
https://gist.github.com/Talkless/28fe39d2176e00b25d9987caef9009be (using gist as log conflicts with github formatting somehow).

Intrestingly, if I run same `cmake` command again (second time) - build succeeds (`openssl` is built)!

If I remove `openssl` package with `conan.exe remove openssl --force` and run `cmake` again, it also succeeds!

Now to actually reproduce build, I have to remove `openssl` package *and* `build` directory.

This issue is experienced only with `openssl` package so far.

For the original issue "it always works the second time", I'd stand by my first suggestion for OP... We have sadly seen this many times.

***

As for the path escape problem... I think I found the issue 🤔 

❓ Are you specifying the compiler?

the following lines prepare the `CC` and `CXX` variables and pass them to perl/msys to apply in the build scripts.

https://github.com/conan-io/conan-center-index/blob/8feafc2df48150492ba13a4d6d09c0d073f33e80/recipes/openssl/1.x.x/conanfile.py#L551-L552

Using a normal command prompt, I get

```txt
openssl/1.1.1j: Calling build()
**********************************************************************
** Visual Studio 2019 Developer Command Prompt v16.7.7
** Copyright (c) 2020 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64'
openssl/1.1.1j: using target: VC-conan-Release-Windows-x86_64-Visual Studio-16 -> VC-WIN64A
openssl/1.1.1j: my %targets = (
    "VC-conan-Release-Windows-x86_64-Visual Studio-16" => {
        inherit_from => [ "VC-WIN64A" ],
        cflags => add("-O2 -Ob2 -MD"),
        cxxflags => add("-O2 -Ob2 -MD"),
        defines => add("NDEBUG"),
        includes => add("C:/.conan/9d4cd9/1/include"),
        lflags => add(""),
    },
);
```

Shortly after it happily compiles

```
Microsoft (R) Program Maintenance Utility Version 14.27.29112.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        cl  /Zi /Fdossl_static.pdb /Gs0 /GF /Gy /MD -O2 -Ob2 -MD /W3 /wd4090 /nologo /O2 /I "." /I "include" -D"L_ENDIAN" -D"OPENSSL_PIC" -D"OPENSSL_CPUID_OBJ" -D"OPENSSL_IA32_SSE2" -D"OPENSSL_BN_ASM_MONT" -D"OPENSSL_BN_ASM_MONT5" -D"OPENSSL_BN_ASM_GF2m" -D"SHA1_ASM" -D"SHA256_ASM" -D"SHA512_ASM" -D"KECCAK1600_ASM" -D"RC4_ASM" -D"MD5_ASM" -D"AESNI_ASM" -D"VPAES_ASM" -D"GHASH_ASM" -D"ECP_NISTZ256_ASM" -D"X25519_ASM" -D"POLY1305_ASM" -D"OPENSSLDIR=\"C:\\Users\\Chris McArthur\\.conan\\data\\openssl\\1.1.1j\\_\\_\\package\\127af201a4cdf8111e2e08540525c245c9b3b99e\\res\"" -D"ENGINESDIR=\"C:\\Users\\Chris McArthur\\.conan\\data\\openssl\\1.1.1j\\_\\_\\package\\127af201a4cdf8111e2e08540525c245c9b3b99e\\lib\\engines-1_1\"" -D"OPENSSL_SYS_WIN32" -D"WIN32_LEAN_AND_MEAN" -D"UNICODE" -D"_UNICODE" -D"_CRT_SECURE_NO_DEPRECATE" -D"_WINSOCK_DEPRECATED_NO_WARNINGS" -D"OPENSSL_USE_APPLINK" -D"NDEBUG" -D"NDEBUG" -I"C:/.conan/9d4cd9/1/include"  -c /Foapps\app_rand.obj "apps\app_rand.c"
```

↪️ Notice it's using `cl.exe` directly from my path, so no funky path problems,

However with `CC` and `CXX` as follows

```env
cc=C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx86/x86/cl.exe
cxx=C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx86/x86/cl.exe
```

OpenSSL build receives

```
openssl/1.1.1j: Calling build()
Conan:vcvars already set
openssl/1.1.1j: using target: VC-conan-Release-Windows-x86_64-Visual Studio-16 -> VC-WIN64A
openssl/1.1.1j: my %targets = (
    "VC-conan-Release-Windows-x86_64-Visual Studio-16" => {
        inherit_from => [ "VC-WIN64A" ],
        cflags => add("-O2 -Ob2 -MD"),
        cxxflags => add("-O2 -Ob2 -MD"),
        defines => add("NDEBUG"),
        includes => add("C:/.conan/9d4cd9/1/include"),
        lflags => add(""),
        cc => "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\bin\Hostx86\x86\cl.exe",
        cxx => "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.27.29110\bin\Hostx86\x86\cl.exe",
    },
);
```

Than I can reproduce the original issues

```
C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx86/x86/cl.exe  /Zi /Fdossl_static.pdb /Gs0 /GF /Gy /MD -O2 -Ob2 -MD /W3 /wd4090 /nologo /O2 /I "." /I "include" -D"L_ENDIAN" -D"OPENSSL_PIC" -D"OPENSSL_CPUID_OBJ" -D"OPENSSL_IA32_SSE2" -D"OPENSSL_BN_ASM_MONT" -D"OPENSSL_BN_ASM_MONT5" -D"OPENSSL_BN_ASM_GF2m" -D"SHA1_ASM" -D"SHA256_ASM" -D"SHA512_ASM" -D"KECCAK1600_ASM" -D"RC4_ASM" -D"MD5_ASM" -D"AESNI_ASM" -D"VPAES_ASM" -D"GHASH_ASM" -D"ECP_NISTZ256_ASM" -D"X25519_ASM" -D"POLY1305_ASM" -D"OPENSSLDIR=\"C:\\Users\\Chris McArthur\\.conan\\data\\openssl\\1.1.1j\\_\\_\\package\\127af201a4cdf8111e2e08540525c245c9b3b99e\\res\"" -D"ENGINESDIR=\"C:\\Users\\Chris McArthur\\.conan\\data\\openssl\\1.1.1j\\_\\_\\package\\127af201a4cdf8111e2e08540525c245c9b3b99e\\lib\\engines-1_1\"" -D"OPENSSL_SYS_WIN32" -D"WIN32_LEAN_AND_MEAN" -D"UNICODE" -D"_UNICODE" -D"_CRT_SECURE_NO_DEPRECATE" -D"_WINSOCK_DEPRECATED_NO_WARNINGS" -D"OPENSSL_USE_APPLINK" -D"NDEBUG" -D"NDEBUG" -I"C:/.conan/9d4cd9/1/include"  -c /Foapps\app_rand.obj "apps\app_rand.c"
```

It does not print the `app_rand.c` as it should when it's successful.

Which is later followed by

```
LINK : fatal error LNK1181: cannot open input file 'apps\app_rand.obj'
```

***

As for fixing the problem, I'm at odds.

Adding `"` around the path gives

```
Bareword found where operator expected at (eval 13) line 13, near """C"
        (Missing operator before C?)

Failure!  build file wasn't produced.
Please read INSTALL and associated NOTES files.  You may also have to look over
your available compiler tool chain or change your configuration.

syntax error at (eval 13) line 13, near """C"
Unknown regexp modifier "/M" at (eval 13) line 13, at end of line
Unknown regexp modifier "/r" at (eval 13) line 13, at end of line
Unknown regexp modifier "/f" at (eval 13) line 13, at end of line
Unknown regexp modifier "/t" at (eval 13) line 13, at end of line
Unknown regexp modifier "/C" at (eval 13) line 13, at end of line
Unknown regexp modifier "/t" at (eval 13) line 13, at end of line
Unknown regexp modifier "/y" at (eval 13) line 13, at end of line
Unknown regexp modifier "/T" at (eval 13) line 13, at end of line
Unknown regexp modifier "/1" at (eval 13) line 13, at end of line
(eval 13) has too many errors.
```

And using `\ ` does not help

```
        cc => "C:/Program\ Files\ (x86)/Microsoft\ Visual\ Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx86/x86/cl.exe",
        cxx => "C:/Program\ Files\ (x86)/Microsoft\ Visual\ Studio/2019/Community/VC/Tools/MSVC/14.27.29110/bin/Hostx86/x86/cl.exe",
```

I still have the original issue 😑 

To my knowledge thus far it seems to be a problem internal with the OpenSSL `nmake` in `perl` that is causing the problem.

--- a/recipes/openssl/1.x.x/conanfile.py
+++ b/recipes/openssl/1.x.x/conanfile.py
@@ -635,6 +635,13 @@ class OpenSSLConan(ConanFile):
             if self._use_nmake and self._full_version >= "1.1.0":
                 self._replace_runtime_in_file(os.path.join("Configurations", "10-main.conf"))
 
+            if self._use_nmake:
+                # Windows: when cmake generates its cache, it populates some environment variables as well.
+                # If cmake also initiates openssl build, their values (containing spaces and forward slashes)
+                # break nmake (don't know about mingw make). So we fix them
+                for v in ['CC', 'CXX', 'RC']:
+                    if v in os.environ and not '"' in os.environ[v]:
+                        os.environ[v] = '"' + os.environ[v].replace('/', '\\') + '"'
             self.run('{perl} ./Configure {args}'.format(perl=self._perl, args=args), win_bash=self._win_bash)
 
             self._patch_install_name()

--- a/recipes/openssl/1.x.x/conanfile.py
+++ b/recipes/openssl/1.x.x/conanfile.py
@@ -1,6 +1,7 @@
 import os
 import fnmatch
 import textwrap
+from contextlib import contextmanager
 from functools import total_ordering
 from conans.errors import ConanInvalidConfiguration
 from conans import ConanFile, AutoToolsBuildEnvironment, tools
@@ -635,13 +636,6 @@ class OpenSSLConan(ConanFile):
             if self._use_nmake and self._full_version >= "1.1.0":
                 self._replace_runtime_in_file(os.path.join("Configurations", "10-main.conf"))
 
-            if self._use_nmake:
-                # Windows: when cmake generates its cache, it populates some environment variables as well.
-                # If cmake also initiates openssl build, their values (containing spaces and forward slashes)
-                # break nmake (don't know about mingw make). So we fix them
-                for v in ['CC', 'CXX', 'RC']:
-                    if v in os.environ and not '"' in os.environ[v]:
-                        os.environ[v] = '"' + os.environ[v].replace('/', '\\') + '"'
             self.run('{perl} ./Configure {args}'.format(perl=self._perl, args=args), win_bash=self._win_bash)
 
             self._patch_install_name()
@@ -687,6 +681,20 @@ class OpenSSLConan(ConanFile):
             return "gcc"
         return "cc"
 
+    @contextmanager
+    def _make_context(self):
+        if self._use_nmake:
+            # Windows: when cmake generates its cache, it populates some environment variables as well.
+            # If cmake also initiates openssl build, their values (containing spaces and forward slashes)
+            # break nmake (don't know about mingw make). So we fix them
+            def sanitize_env_var(var):
+                return '"{}"'.format(var).replace('/', '\\') if '"' not in var else var
+            env = {key: sanitize_env_var(tools.get_env(key)) for key in ("CC", "RC") if tools.get_env(key)}
+            with tools.environment_append(env):
+                yield
+        else:
+            yield
+
     def build(self):
         with tools.vcvars(self.settings) if self._use_nmake else tools.no_op():
             env_vars = {"PERL": self._perl}
@@ -706,7 +714,8 @@ class OpenSSLConan(ConanFile):
                 else:
                     self._patch_configure()
                     self._patch_makefile_org()
-                self._make()
+                with self._make_context():
+                    self._make()
 
     @property
     def _cross_building(self):

